frontend/src/ChatMessage.tsx

```diff
@@ -41,7 +46,16 @@ interface IProps {
   message: IMessage;
 }
 const ChatMessage = (props: IProps): JSX.Element => {
+  const [fireworkStarted, setFireworkStarted] = React.useState<boolean>(false);
```

Comment:
Can we factor the `firework` into its own component, similar to the Beeper? Then, we can lazily include the firework on the associated with the winner announcement, and the firework can manage its own state.

Reply:
Yes. we can do that. The situation for Beeper is quite different from firework though. firework is a response to an event, whereas Beeper is for config setting. To properly handle server generated events, we will eventually need to have some kind of event handling capacities on both client and server side. ChatMessage is for rendering the messages. It actually causes firework to fire each time the chat box renders its messages without hacking the rendering. Since the underlying mechanism is essentially a state, hacking it to handle events will not look that great in the long run.

Reply:
There are event handling capabilities in the websocket handler -- it probably wouldn't be difficult for that to create a react event for a hook. But, modeling the firework as a component that renders fireworks while it is rendered should make this moot? IIUC idiomatic React prefers to minimize "events" in favor of state-rendering

Reply:
There is a difference in moving the firework to the component three. The current hack ensures each winning message generates one and only one firework by associating firework with each winning message (an instance of ChatMessage). When another winning message arrives, it will trigger another firework (from another instance of ChatMessage). When we create a firework component in the component tree, simply duplicating the logic will have the effect of generating one firework regardless of number of winning messages received. If we use state to handle when to trigger firework, then we are back to where we were before this in previous branch. The firework component will need to be able to somehow generate the firework exactly once for each winning message. 

question to you, if I use the key attribute with different value, will react create a different component instance even though the component is not part of a list? Thanks.

Reply:
> question to you, if I use the key attribute with different value, will react create a different component instance even though the component is not part of a list? Thanks.

Honestly, not sure. But hopefully not relevant?

> the current hack ...

What I am thinking is to have the component set up the firework animation (including setInterval, etc) in `useEffect`, and then use the function-return-value from `useEffect` to implement tearDown.

Something like this:

```
const Firework = (props) => {
  const { setInterval, clearInterval } = React.useContext(TimerContext);
  const { elapsed, setElapsed } = React.useState(0);
  React.useEffect(() => {
     // set up handlers
    const t = setInterval(() => {
      // ...
      setElapsed(Date.now());
    }, 100);
    return () => {
      clearInterval(t);
      // clean up other handlers
    }
  }, [])

  return (<>
    <canvas...>
  </>);
}
```

Reply:
See my changes in the upcoming commit. Don't think the same approach as Beeper works since it does not resolve the issue of when to trigger the fireworks.

The Firework component is just a simple wrap to make ChatMessage cleaner, and still adhere to the semantics of event handling.

Also, no clean up is needed for confetti as it cleans up by itself, and the instance can be reused when startFireworks is called again (if ever)

Reply:
https://www.npmjs.com/package/react-confetti-canvas

It looks like other folks have already implemented this :)

Reply:
I did see that. Yeah, don't think you will like it. 

It is 3.7M unpacked vs 61.1k.

https://www.npmjs.com/package/canvas-confetti

and weekly downloads are 100+ times more for the non-react version :)

Reply:
I think the vast majority of the size there is the gif in the package preview. The actual source is pretty short.

https://github.com/fufu70/react-confetti-canvas/blob/master/assets/canvas.gif <-- this thing

Anyways, I can poke at this in the evening at the end of the work day.

Reply:
sounds good. The main feature I like about the non-react version is the ability to put the confetti all over the page instead of only in a separate panel. The firework code actually comes from the author's very cool demo page.

https://www.kirilv.com/canvas-confetti/

check it out if you can